### ISSUE TITLE
## Phase 1 Implementation Complete: Dual Embeddings Architecture

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#344]
Successfully completed Tasks #68, #69, and #70 - the foundation for dual parallel embeddings.

### What Was Built

**1. DualRootEmbeddings Class** (Task #68)
- New file: `klareco/embeddings/dual_root_embeddings.py`
- Two independent 64d embeddings (linguistic + topical)
- Three modes: 'linguistic', 'topical', 'combined' (128d)
- Methods: forward(), get_normalized(), similarity()
- Freezing/unfreezing support for sequential training
- Load linguistic from RootEmbeddings checkpoint
- 19 comprehensive unit tests (all passing)

**2. Topical Data Preparation Script** (Task #70)
- New file: `scripts/data/prepare_topical_pairs.py`
- Shell wrapper: `scripts/prepare_topical_data.sh`
- Extracts skip-gram pairs from 4.3M corpus
- Window size=5, negative ratio=5:1
- Checkpointing support (resumable)
- Function word filtering (prevents collapse)
- Graded similarity targets based on distance
- Tested with 1000 sentences: generated 104K pairs

**3. CompositionalEmbedding Integration** (Task #69)
- Modified: `klareco/embeddings/compositional.py`
- Added `use_dual_roots` parameter (default=False)
- Backward compatible with existing code
- Methods: set_root_embedding_mode(), load_linguistic_embeddings()
- Freeze/unfreeze methods for training control
- Fixed device compatibility issue
- 14 comprehensive unit tests (all passing)

### Key Design Decisions

1. **Sequential Training Recommended**: Train linguistic first, freeze, then train topical
   - Safer than joint training (avoids interference)
   - Can reuse existing linguistic embeddings
   - Faster convergence

2. **Skip-gram Window Size = 5**: Captures medium-range context
   - Distance 1: target=0.7 (very similar)
   - Distance 5: target=0.3 (weakly related)

3. **Negative Sampling 5:1**: Standard ratio for word2vec-style training

4. **Backward Compatibility**: All existing code works unchanged
   - Default `use_dual_roots=False` preserves single embedding behavior
   - Can load old checkpoints

### Next Steps (Phase 2 - Week 2)

**Task #71**: Implement dual training script
- Create `scripts/training/train_dual_embeddings.py`
- Two-stage training: linguistic (reuse existing) + topical (new)
- Multi-task loss: MSE + contrastive
- Depends on: #68, #70 (both complete)

**Before #71**: Generate full topical training data
```bash
./scripts/prepare_topical_data.sh --resume
# Will process 4.3M sentences, ~50M pairs
# Takes ~2-4 hours, checkpointed every 500K sentences
```

### Files Created
- `klareco/embeddings/dual_root_embeddings.py` (269 lines)
- `klareco/embeddings/__init__.py` (updated)
- `scripts/data/prepare_topical_pairs.py` (518 lines)
- `scripts/prepare_topical_data.sh` (shell wrapper)
- `tests/test_dual_root_embeddings.py` (338 lines, 19 tests)
- `tests/test_compositional_dual_roots.py` (335 lines, 14 tests)

### Files Modified
- `klareco/embeddings/compositional.py` (+90 lines, 6 new methods)

### Test Coverage
- DualRootEmbeddings: 19/19 tests passing
- CompositionalEmbedding: 14/14 tests passing
- Total: 33 tests, 100% passing

### Time Spent
- Task #68: ~4 hours (implementation + tests)
- Task #70: ~3 hours (script + testing)
- Task #69: ~2 hours (integration + tests)
- **Total: ~9 hours** (on schedule for 12-hour Phase 1 estimate)

**Comment by marctjones** [comment]
Closing session summary - knowledge consolidated into wiki and CURRENT_PLAN.md
